esp/bindings/http/platform/httpservice.cpp

```diff
@@ -974,3 +875,552 @@ int CEspHttpServer::onGet()
+        //Create a new session and redirect to the login page.
+        askUserLogin(authReq.authBinding, createHTTPSession(authReq, nullptr, "/"));
+        return authSucceeded;
+    }
+
+    if (sessionID > 0)
+        return doSessionAuth(sessionID, authReq, nullptr, nullptr);//Check session based authentication using this session ID.
+
+    if(!readCookie(SESSION_AUTH_COOKIE))
+        return authUnknown;
+
+    //The session has expired.
+    //If userName/password are received (from login page), we still can check session based authentication.
+    const char* userName = (authReq.requestParams) ? authReq.requestParams->queryProp("username") : NULL;
+    const char* password = (authReq.requestParams) ? authReq.requestParams->queryProp("password") : NULL;
+    if (!isEmptyString(userName) && !isEmptyString(password))
+        return doSessionAuth(createHTTPSession(authReq, nullptr, "/"), authReq, userName, password);
+
+    if(authReq.isSoapPost) //from SOAP Test page
+        sendMessage("Session expired. Please close this page and login again.", "text/html; charset=UTF-8");
+    else //from other page
+        askUserLogin(authReq.authBinding, createHTTPSession(authReq, nullptr, nullptr));
+    return authFailed;
+}
+
+EspAuthState CEspHttpServer::checkUserAuthPerRequest(EspAuthRequest& authReq)
+{
+    ESPLOG(LogMax, "checkUserAuthPerRequest");
+
+    authReq.authBinding->populateRequest(m_request.get());
+    if (authReq.authBinding->doAuth(authReq.ctx))
+    {//We do pass the authentication per the request
+        // authenticate optional groups. Do we still need?
+        authOptionalGroups(authReq);
+
+        StringBuffer userName, peer;
+        ESPLOG(LogNormal, "Authenticated for %s@%s", authReq.ctx->getUserID(userName).str(), m_request->getPeer(peer).str());
+        return authSucceeded;
+    }
+    if(!authReq.isSoapPost)
+        return authUnknown;
+
+    //If SoapPost, username/password may be in soap:Header which is not in HTTP header.
+    //The doAuth() may check them inside CSoapService::processHeader() later.
+    authReq.ctx->setToBeAuthenticated(true);
+    return authPending;
+}
+
+void CEspHttpServer::sendMessage(const char* msg, const char* msgType)
+{
+    m_response->setContent(msg);
+    m_response->setContentType(msgType);
+    m_response->setStatus(HTTP_STATUS_OK);
+    m_response->send();
+}
+
+EspAuthState CEspHttpServer::doSessionAuth(unsigned sessionID, EspAuthRequest& authReq, const char* _userName, const char* _password)
+{
+    ESPLOG(LogMax, "doSessionAuth: %s<%u>", PropSessionID, sessionID);
+
+    CDateTime now;
+    now.setNow();
+    time_t timeNow = now.getSimple();
+    Owned<IRemoteConnection> conn = querySDSConnection(authReq.authBinding->queryDomainSessionSDSPath(), RTM_LOCK_WRITE, SESSION_SDS_LOCK_TIMEOUT);
+    IPropertyTree* domainSessions = readAndCleanDomainSessions(authReq.authBinding, conn, timeNow);
```

Comment:
Also, it is better where possible to pass in IPT's rather than IRemoteConnection's, as it's more portable if tree comes from somewhere else. 
And, if you don't need the functionality of a connection in a method, you shouldn't really pass it (and I don't think you do in readAndCleanDomainSessions).

Pass conn->queryRoot() instead (after validating conn if necessary).